Add documentation for Docker deploys. (#465)
vapor/docs: 4.0/docs/deploy/docker.md

# Deploying with Docker
Using Docker to deploy your Vapor app has several benefits: 

1. Your dockerized app can be spun up reliably using the same commands on any platform with a Docker Daemon written for it -- namely, Linux (CentOS, Debian, Fedora, Ubuntu), macOS, and Windows.
2. You can use docker-compose or Kubernetes manifests to coordinate launching of your app with launching of other services needed for a full deployment (e.g. Redis, Postgres, nginx, etc.).
3. It is easy to test your app's ability to scale horizontally, even locally on your development machine.

This guide will stop short of explaining how to get your dockerized app onto a server. The simplest deploy would involve installing Docker on your server and running the same commands you would run on your development machine to spin things up remotely. 

> More complicated and robust deployments are usually different depending on your hosting solution; many popular solutions like AWS have builtin support for Kubernetes and custom database solutions which make it difficult to write best practices in a way that applies to all deployments. 

Reviewer: I think it makes sense to explain the limitations of this guide at the beginning as matt is doing here.

The docker-compose file in your new Vapor App will define services for running your app, running migrations or reverting them, and running a database as your app's persistence layer. The exact definitions will vary depending on which database you chose to use when you ran `vapor-beta new`.

Note that your docker-compose file has some shared environment variables near the top.
```docker
x-shared_environment: &shared_environment
  LOG_LEVEL: ${LOG_LEVEL:-debug}
  DATABASE_HOST: db
  DATABASE_NAME: vapor_database
  DATABASE_USERNAME: vapor_username
  DATABASE_PASSWORD: vapor_password
```

You will see these pulled into multiple services below with the `<<: *shared_environment` YAML reference syntax.

The `DATABASE_HOST`, `DATABASE_NAME`, `DATABASE_USERNAME`, and `DATABASE_PASSWORD` variables are hard coded in this example whereas the `LOG_LEVEL` will take its value from the environment running the service or fall back to `'debug'` if that variable is unset.

!!! note

>     Hard-coding the username and password for database access is one of a number of areas where this docker-compose file sets itself up for development use but does _not_ represent best-practices for production deployment.

Author: The problem is that a production docker-compose file would be much more complicated (in my experience, production uses involve two or more docker-compose files to provide a base file and additionally overrides for local development) and this docker-compose file is actually just fine for local development purposes.

I think I can word this a bit more positively, though.

>     Hard-coding the username and password for database access is one of a number of areas where this docker-compose file sets itself up for development use but does _not_ represent best-practices for production deployment.

Reviewer 2: Sounds good - I agree, I think in this case just rewording would be enough

First, execute
```shell
docker-compose up app
```
and notice that both the `app` and `db` services are started.

Your app is listening on port 80 _inside the Docker container_ but as defined by the docker-compose file, it is accessible on your developer machine at **http://localhost:8080**. 

This port mapping distinction is very important because you can run any number of services on the same ports if they are all running in their own containers and they each expose different ports to the host machine.

Visit `http://localhost:8080` and you will see `It works!` but visit `http://localhost:8080/todos` and you will get: 
```
{"error":true,"reason":"Something went wrong."}
```

Take a peak at the logs output in the terminal where you ran `docker-compose up app` and you will see
```
[ ERROR ] relation "todos" does not exist
```

Of course! We need to run migrations on the database. Press `Ctrl+C` to bring your app down. We are going to start the app up again but this time with
```shell
docker-compose up --detach app
```

> Now your app is going to start up "detached" (in the background). You can prove to yourself that your app is running with

Reviewer 2: ```suggestion
Now your app is going to start up "detached" (in the background). You can verify this by running
```

```
but the easiest way to bring these particular containers down is
```shell
docker-compose down
```

#### Wiping your database
The docker-compose file defines a `db_data` volume to persist your database between runs. There are a couple of ways to reset your database.

You can remove the `db_data` volume at the same time as bringing your containers down with
```shell
docker-compose down --volumes
```

You can see any volumes currently persisting data with `docker volume ls`. Note that the volume name will generally have a prefix of `my-dockerized-app_` or `test_` depending on whether you were running in Swarm Mode or not. 

You can remove these volumes one at a time with e.g.
```shell
docker volume rm my-dockerized-app_db_data
```

You can also clean up all volumes with
```shell
docker volume prune
```

Just be careful you don't accidentally prune a volume with data you wanted to keep around!

Docker will not let you remove volumes that are currently in use by running or stopped containers. You can get a list of running containers with `docker container ls` and you can see stopped containers as well with `docker container ls -a`.

### Swarm Mode

> In addition to being "the production orchestrator that lost to Kubernetes," Swarm Mode is an easy interface to use when you've got a docker-compose file handy and you want to test how your app scales horizontally. You can read all about Swarm Mode in the pages rooted at the [overview](https://docs.docker.com/engine/swarm/).

Reviewer 2: ```suggestion
Swarm Mode is an easy interface to use when you've got a docker-compose file handy and you want to test how your app scales horizontally. You can read all about Swarm Mode in the pages rooted at the [overview](https://docs.docker.com/engine/swarm/).
```

# Deploying with Docker

> Using Docker to deploy your Vapor app has several benefits: 

Reviewer: Nit: newline after headers (for whole doc)

2. You can use docker-compose or Kubernetes manifests to orchestrate multiple services needed for a full deployment (e.g. Redis, Postgres, nginx, etc.).
3. It is easy to test your app's ability to scale horizontally, even locally on your development machine.

This guide will stop short of explaining how to get your dockerized app onto a server. The simplest deploy would involve installing Docker on your server and running the same commands you would run on your development machine to spin up your application. 

More complicated and robust deployments are usually different depending on your hosting solution; many popular solutions like AWS have builtin support for Kubernetes and custom database solutions which make it difficult to write best practices in a way that applies to all deployments. 

Nevertheless, using Docker to spin your entire server stack up locally for testing purposes is incredibly valuable for both big and small serverside apps. Additionally, the concepts described in this guide apply in broad strokes to all Docker deployments.

> ## Getting set up

Reviewer: Is there anything we could put after this header? I try to avoid stacking multiple headers directly since it can look a little cluttered.

Reviewer: nit: Headers should use title capitalization

### Install Docker
You will need to install Docker for your developer environment. You can find information for any platform in the [Supported Platforms](https://docs.docker.com/install/#supported-platforms) section of the Docker Engine Overview. If you are on Mac OS, you can jump straight to the [Docker for Mac](https://docs.docker.com/docker-for-mac/install/) install page.

> ### Generate a dockerized Vapor App

Reviewer: Any way to make this more concise? Headers can be used to generate links and we should consider how they will look in this context. For example, this would be:

```
docs.vapor.codes/4.0/deploy/docker#generate-a-dockerized-vapor-app
```

- `vapor` is redundant here
- `generate` seems superfluous 

Maybe this would be better:

```
docs.vapor.codes/4.0/deploy/docker#dockerized-app
```

or maybe `#dockerize` or `#docker-template`?

FROM vapor/swift:5.2 as build
WORKDIR /build

# Copy entire repo into container
COPY . .

# Compile with optimizations
RUN swift build \
	--enable-test-discovery \
	-c release \
	-Xswiftc -g
```

Next, it pulls in a different image (`vapor/ubuntu:18.04`) as a second stage. The second stage copies what is needed to run your app and omits build-only dependencies. 
```docker
# ================================
# Run image
# ================================
FROM vapor/ubuntu:18.04
WORKDIR /run

# Copy build artifacts
COPY --from=build /build/.build/release /run
# Copy Swift runtime libraries
COPY --from=build /usr/lib/swift/ /usr/lib/swift/
# Uncomment the next line if you need to load resources from the `Public` directory
#COPY --from=build /build/Public /run/Public
```

Finally, the Dockerfile defines an entrypoint and a default command to run.
```docker
ENTRYPOINT ["./Run"]
CMD ["serve", "--env", "production", "--hostname", "0.0.0.0"]
```

Both of these can be overridden when the image is used, but by default running your image will result in calling the `serve` command of your app's `Run` target. The full command on launch will be `./Run serve --env production --hostname 0.0.0.0`.

> ### docker-compose

Reviewer: This should at least be `Docker Compose` for the title (if not everywhere).
